internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,365 @@
+
+// Reconcile reconciles a CAPI cluster, creating a Rancher cluster if needed and applying the import manifests.
+func (r *CAPIImportManagementV3Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling CAPI cluster")
+
+	capiCluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, capiCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	patchBase := client.MergeFromWithOptions(capiCluster.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+	log = log.WithValues("cluster", capiCluster.Name)
+
+	// Wait for controlplane to be ready. This should never be false as the predicates
+	// do the filtering.
+	if !capiCluster.Status.ControlPlaneReady && !conditions.IsTrue(capiCluster, clusterv1.ControlPlaneReadyCondition) {
+		log.Info("clusters control plane is not ready, requeue")
+		return ctrl.Result{RequeueAfter: defaultRequeueDuration}, nil
+	}
+
+	// Collect errors as an aggregate to return together after all patches have been performed.
+	var errs []error
+
+	result, err := r.reconcile(ctx, capiCluster)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error reconciling cluster: %w", err))
+	}
+
+	if err := r.Client.Patch(ctx, capiCluster, patchBase); err != nil {
+		errs = append(errs, fmt.Errorf("failed to patch cluster: %w", err))
+	}
+
+	if len(errs) > 0 {
+		return ctrl.Result{}, errorutils.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcile(ctx context.Context, capiCluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// placeholder name/namespace must be set to be able to test this but it won't be used.
+	rancherCluster := &managementv3.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "c-",
+			Namespace: capiCluster.Namespace,
+			Labels: map[string]string{
+				capiClusterOwner:          capiCluster.Name,
+				capiClusterOwnerNamespace: capiCluster.Namespace,
+			},
+		},
+	}
+
+	rancherClusterList := &managementv3.ClusterList{}
+	selectors := []client.ListOption{
+		client.MatchingLabels{
+			capiClusterOwner:          capiCluster.Name,
+			capiClusterOwnerNamespace: capiCluster.Namespace,
+			ownedLabelName:            "",
+		},
+	}
+	err := r.RancherClient.List(ctx, rancherClusterList, selectors...)
+
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, fmt.Sprintf("Unable to fetch rancher cluster %s", client.ObjectKeyFromObject(rancherCluster)))
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if len(rancherClusterList.Items) != 0 {
+		if len(rancherClusterList.Items) > 1 {
+			log.Info("More than one rancher cluster found. Will default to using the first one.")
+		}
+
+		rancherCluster = &rancherClusterList.Items[0]
+	}
+
+	if !capiCluster.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
Do we set a finalizer on the CAPI Cluster object to prevent situations where we miss the event of cluster removal, so the Rancher v3.Cluster is left orphaned?

Reply:
No we don't and thats a good point Danil. 

@salasberryfin - would you be able to add a finalizer as part of the reconcile.

Reply:
Thanks, I'll be fixing this

Reply:
Last version adds a finalizer to the capi cluster (only if the cluster is not annotated) which is deleted whenever the rancher cluster is removed.